rfcs/0077-layout-transform-padding.md

```diff
@@ -0,0 +1,3090 @@
+          # These later accesses of B[0] may not be replaced, because
+          # for all loop iterations i!=0, the value stored in B[0] has
+          # been overwritten since the T.assume call.
+          for i in T.serial(16):
+              B[0] = B[0] + A[i]
+  ```
+
+### New Transform - Hoist Expression
+
+A new utility `HoistExpression`, which is a generalization of the
+current `HoistIfThenElse` pass.  The transformation `HoistExpression`
+would apply to the entire body of the `PrimFunc`, and would be used to
+avoid duplication of functionality between `HoistIfThenElse` and
+`HoistExpression`.
+
+`HoistExpression` would also be exposed as a metaschedule primitive,
+acting within a specified block of the `PrimFunc`, with the
+configuration options given below.
+
+```c++
+enum class HoistConditional {
+  kNone = 0,
+  kIfElseStmt = (1<<0),
+  kIfElseExpr = (1<<1),
+  kBooleanExpression = (1<<2),
+};
+
+enum class HoistLetBinding {
+  kNone = 0,
+  kRequiredByCondition = (1<<0),
+  kLetStmt = (1<<1),
+  kLetExpr = (1m<<2),
+};
+```
+
+* The values in `HoistConditional` are bit flags, indicating which
+  conditionals should be hoisted.
+
+  * `HoistConditional::kNone` - Do not hoist conditionals
+
+  * `HoistConditional::kIfElseStmt` - If set, attempt to hoist
+    conditionals that occur within `IfThenElseNode::condition`.
+
+  * `HoistConditional::kIfElseExpr` - If set, attempt to hoist
+    conditionals that occur as the condition of a
+    `builtin::if_then_else` call.
+
+  * `HoistConditional::kBooleanExpression` - If set, attempt to hoist
+    any `PrimExpr` whose data type is `DataType::Bool()`.
+
+* The values in `HoistLetBindings` are bit flags, indicating which
+  bindings should be hoisted.
+
+  * `HoistLetBinding::kNone` - Do not hoist any let bindings.
+
+  * `HoistLetBinding::kRequiredByCondition` - If set, hoist a let
+    binding if it is required in order to hoist a conditional.
+
+  * `HoistLetBinding::kLetStmt = (1<<1)` - If set, attempt to hoist
+    any let bindings performed using `LetStmt`.
+
+  * `HoistLetBinding::kLetExpr` - If set, attempt to hoist any let
+    bindings performed using `Let`.
+
+The existing pass `HoistIfElse` is roughly equivalent to using
+`HoistExpression` with `HoistConditional::kIfElseStmt` and
+`HoistLetBinding::kNone`.  The one exception is that `HoistIfElse`
+occurs after all let bindings have been inlined, and does not check
+let bindings when determining if a condition can be hoisted.
+
+```python
+# Original function
+@T.prim_func
+def func(A: T.Buffer[(4,4), "float32"]):
+    for i in T.serial(4):
+        is_in_bounds = i < 3
+        if is_in_bounds:
+            A[i] = 0.0
+
+# Incorrectly hoisted by `HoistIfThenElse`
+@T.prim_func
+def func(A: T.Buffer[(4,), "float32"]) -> None:
+    is_in_bounds = T.var("bool")
+    if is_in_bounds:
+        for i in T.serial(4):
+            is_in_bounds = i < 3
+            A[i] = 0.0
+```
+
+### New Transform - Reduce Loop Extents
```

Comment:
That's a good point, and would avoid having the special case pass.

Looking at the implementation of `LoopPartition` for edge cases, and making some notes to myself on the steps required.

* Apply a `T.likely` annotation when checking if it's a pad value (e.g. `if !T.likely(4*io + ii < 14)`), since `LoopPartition` uses this to identify partitionable conditions.
* Maintain the `T.likely` annotation when hoisting part of a conditional (e.g. When hoisting `io==3` out of `!T.likely(io==3 and ii >=2`).
* Look into relaxing the restriction against partitioning a constant loop, currently only allowed by a pass config.  If we're generating loops that we know should be partitioned, it would be strange to also require the user to opt-in.  I don't know the history of this restriction, so this would require some investigation.  (Perhaps could allow the additional partition only if the loop is a serial loop and all but one of the partitions are no-ops.)